lib/index.js

```diff
@@ -179,32 +184,36 @@ export function configureUnifiedLanguageServer(
   async function processDocuments(textDocuments, alwaysStringify = false) {
     // LSP uses `file:` URLs (hrefs), `unified-engine` expects a paths.
     // `process.cwd()` does not add a final slash, but `file:` URLs often do.
-    const workspacesAsPaths = [...workspaces].map((d) =>
-      fileURLToPath(d.replace(/\/$/, ''))
-    )
+    const workspacesAsPaths = [...workspaces].map((d) => d.replace(/\/?$/, '/'))
     /** @type {Map<string, Array<VFile>>} */
     const workspacePathToFiles = new Map()
 
-    if (workspacesAsPaths.length === 0) {
-      workspacesAsPaths.push(process.cwd())
-    }
+    await Promise.all(
+      textDocuments.map(async (textDocument) => {
+        let [cwd] = workspacesAsPaths
+          // Every workspace that includes the document.
+          .filter(
+            (d) => textDocument.uri.slice(0, d.length + 1) === d + path.sep
+          )
+          // Sort the longest (closest to the file) first.
+          .sort((a, b) => b.length - a.length)
+        if (!cwd) {
+          cwd = await packageDirectory({
+            cwd: path.dirname(fileURLToPath(textDocument.uri))
+          })
+        }
```

Comment:
I recommend switching between the 2 behaviors based on `workspacesAsPaths.length > 0`, instead of using one as a fallback for both o workspaces existing and a file not in a workspace.

Seems easier to reason about and debug